klippy/extras/axis_twist_compensation.py

```diff
@@ -0,0 +1,347 @@
+            interval_dist = \
+                (self.end_x - self.start_x) / (n_points - 1)
+            indexes = [self.start_x + i*interval_dist
+                       for i in range(0, n_points)]
+
+            # Calculate the mean of x and y values
+            mean_indexes = sum(indexes) / n_points
+            mean_z_compensations = sum(z_compensations) / n_points
+
+            # Calculate the covariance and variance
+            covar = sum((indexes[i] - mean_indexes) *
+                        (z_compensations[i] - mean_z_compensations)
+                        for i in range(n_points))
+            var = sum((indexes[i] - mean_indexes)**2 for i in range(n_points))
+
+            # Compute the slope (m) and intercept (b) of the best-fit line
+            self.m = covar / var
+            self.b = mean_z_compensations - self.m * mean_indexes
+
+    def clear_compensations(self):
+        self.z_compensations = []
+        self.m = None
+        self.b = None
+
+    def _register_gcode_handlers(self):
+        self.gcode.register_command(
+            'AXIS_TWIST_COMPENSATION_CLEAR',
+            self.cmd_AXIS_TWIST_COMPENSATION_CLEAR,
+            desc=self.cmd_AXIS_TWIST_COMPENSATION_CLEAR_help)
+
+    cmd_AXIS_TWIST_COMPENSATION_CLEAR_help = \
+        "Clears the active axis twist compensation"
+
+    def cmd_AXIS_TWIST_COMPENSATION_CLEAR(self, gcmd):
+        self.clear_compensations()
+
+
+class Calibrater:
+    def __init__(self, compensation):
+        # setup self attributes
+        self.compensation = compensation
+        self.printer = compensation.printer
+        self.config = compensation.config
+        self.gcode = self.printer.lookup_object('gcode')
+        self.probe = None
+        # probe settings are set to none, until they are available
+        self.lift_speed, self.probe_x_offset, self.probe_y_offset, _ = \
+            None, None, None, None
+        self.printer.register_event_handler("klippy:connect",
+                                            self._handle_connect())
+        self.speed = compensation.speed
+        self.horizontal_move_z = compensation.horizontal_move_z
+        self.start_point = (compensation.start_x, compensation.y)
+        self.end_point = (compensation.end_x, compensation.y)
+        self.results = None
+        self.current_point_index = None
+        self.gcmd = None
+
+        # register gcode handlers
+        self._register_gcode_handlers()
+
+    def _handle_connect(self):
+        # gets probe settings when they are available
+        def callback():
+            self.probe = self.printer.lookup_object('probe', None)
+            if (self.probe is None):
+                raise self.config.error(
+                    "AXIS_TWIST_COMPENSATION requires [probe] to be defined")
+            self.lift_speed = self.probe.get_lift_speed()
+            self.probe_x_offset, self.probe_y_offset, _ = \
+                self.probe.get_offsets()
+        return callback
```

Comment:
It is confusing for `_handle_connect()` to be a factory function - just pass `self._handle_connect` to `self.printer.register_event_handler()` and directly perform the initialization code.  (As is common in all the other Klippy modules.)